contrib/endpoints/src/api_manager/firebase_rules/firebase_request.cc

```diff
@@ -0,0 +1,402 @@
+  auto *fields = variables->mutable_fields();
+
+  path.set_string_value(context_->request()->GetRequestPath());
+  (*fields)[kPath] = path;
+
+  method.set_string_value(
+      GetOperation(context_->request()->GetRequestHTTPMethod()));
+  (*fields)[kMethod] = method;
+
+  SetProtoValue(kToken, claims, &token);
+  (*fields)[kAuth] = token;
+
+  for (auto func_call : function_calls) {
+    status = AddFunctionMock(&request, func_call);
+    if (!status.ok()) {
+      return status;
+    }
+  }
+
+  std::string body;
+  status = utils::ProtoToJson(request, &body, utils::JsonOptions::DEFAULT);
+  if (status.ok()) {
+    env_->LogDebug(std::string("FIREBASE REQUEST BODY = ") + body);
+    firebase_http_request_.body = body;
+  }
+
+  return status;
+}
+
+Status FirebaseRequest::ProcessTestRulesetResponse(const std::string &body) {
+  Status status = utils::JsonToProto(body, &response_);
+  if (!status.ok()) {
+    return status;
+  }
+
+  // If the state is SUCCESS, then we don't need to do any further processing.
+  if (response_.test_results(0).state() ==
+      TestRulesetResponse::TestResult::SUCCESS) {
+    is_done_ = true;
+    next_request_ = nullptr;
+    return Status::OK;
+  }
+
+  // Check that the test results size is 1 since we always send a single test
+  // case.
+  if (response_.test_results_size() != 1) {
+    std::ostringstream oss;
+    oss << "Received TestResultsetResponse with size = "
+        << response_.test_results_size() << " expecting only 1 test result";
+
+    env_->LogError(oss.str());
+    return Status(Code::INTERNAL, "Unexpected TestResultsetResponse");
+  }
+
+  bool allFunctionsProcessed = true;
+
+  // Iterate over all the function calls and make sure that the function calls
+  // are well formed.
+  for (auto func_call : response_.test_results(0).function_calls()) {
+    status = CheckFuncCallArgs(func_call);
+    if (!status.ok()) {
+      return status;
+    }
+    allFunctionsProcessed &= Find(func_call) != funcs_with_result_.end();
+  }
+
+  // Since all the functions have a response and the state is FAILURE, this
+  // means Unauthorized access to the resource.
+  if (allFunctionsProcessed) {
+    std::string message = "Unauthorized Access";
+    if (response_.test_results(0).debug_messages_size() > 0) {
+      std::ostringstream oss;
+      for (std::string msg : response_.test_results(0).debug_messages()) {
+        oss << msg << " ";
+      }
+      message = oss.str();
+    }
+
+    return Status(Code::PERMISSION_DENIED, message);
+  }
+
+  func_call_iter_ = response_.test_results(0).function_calls().begin();
+  return Status::OK;
+}
+
+std::vector<std::pair<FunctionCall, std::string>>::const_iterator
+FirebaseRequest::Find(const FunctionCall &func_call) {
+  return std::find_if(funcs_with_result_.begin(), funcs_with_result_.end(),
+                      [func_call](std::tuple<FunctionCall, std::string> item) {
+                        return MessageDifferencer::Equals(std::get<0>(item),
```

Comment:
As we discussed, creating an string and then comparing is better than just comparing. When ever we get a TestRulesetResponse, creating a serialized String to generate the string key and them trying to do the comparison will be more slower.